src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Agent.Worker.Release.Artifacts.Definition;
+
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Client;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.ReleaseManagement.WebApi.Contracts;
+
+using Newtonsoft.Json;
+
+using ServerBuildArtifact = Microsoft.TeamFoundation.Build.WebApi.BuildArtifact;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts
+{
+    // TODO: Write tests for this
+    public class BuildArtifact : AgentService, IArtifactExtension
+    {
+        public Type ExtensionType => typeof(IArtifactExtension);
+        public AgentArtifactType ArtifactType => AgentArtifactType.Build;
+
+        private const string AllArtifacts = "*";
+
+        public async Task DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, string localFolderPath)
+        {
+            ArgUtil.NotNull(artifactDefinition, nameof(artifactDefinition));
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNullOrEmpty(localFolderPath, nameof(localFolderPath));
+
+            int buildId = Convert.ToInt32(artifactDefinition.Version, CultureInfo.InvariantCulture);
+            if (buildId <= 0)
+            {
+                throw new ArgumentException("artifactDefinition.Version");
+            }
+
+            var buildArtifactDetails = artifactDefinition.Details as BuildArtifactDetails;
+            if (buildArtifactDetails == null)
+            {
+                throw new ArgumentException("artifactDefinition.Details");
+            }
+
+            // Get the list of available artifacts from build. 
+            executionContext.Output(StringUtil.Loc("RMPreparingToGetBuildArtifactList"));
+
+            var vssConnection = new VssConnection(buildArtifactDetails.TfsUrl, buildArtifactDetails.Credentials);
+            var buildClient = vssConnection.GetClient<BuildHttpClient>();
+            var xamlBuildClient = vssConnection.GetClient<XamlBuildHttpClient>();
+            List<ServerBuildArtifact> buildArtifacts = null;
+            DefinitionType buildDefinitionType = DefinitionType.Build;
+
+            try
+            {
+                buildArtifacts = await buildClient.GetArtifactsAsync(buildArtifactDetails.Project, buildId);
+            }
+            catch (BuildNotFoundException)
```

Comment:
yep, thats only one exception we want to handle to get xaml client if build client does not exists